users/vieting/experiments/switchboard/hybrid/feat/network_helpers/perturbation.py

```diff
@@ -0,0 +1,87 @@
+import functools
+import numpy as np
+import torch
+import torchaudio
+
+
+class PerturbationFactor:
+  """
+  Class to wrap perturbation factors, e.g. for speed or tempo perturbation.
+  """
+  def __init__(self, prob, minimum, maximum):
+    self.prob = prob
+    self.min = minimum
+    self.max = maximum
+
+
+class WaveformPerturbation:
+  """
+  Helper class to perform perturbation techniques on audio waveforms.
+  """
+  def __init__(self, speed=None, tempo=None, sox_effects=None, codecs=None, preemphasis=None):
+    """
+    :param speed:
+    """
+    from functools import partial
+    self._speed = PerturbationFactor(**speed) if speed else None
+    self._tempo = PerturbationFactor(**tempo) if tempo else None
+    self._perturbations = [functools.partial(self.sox, sox_effects=sox_effects)]
+    if preemphasis:
+      self._perturbations.append(functools.partial(self.preemphasis, factor=PerturbationFactor(**preemphasis)))
+    if codecs:
+      self._perturbations.append(functools.partial(self.apply_codecs, codecs=codecs))
+
+  def run(self, audio, sample_rate, random_state):
+    # input_shape = audio.shape
+    audio = torch.from_numpy(audio).unsqueeze(0).to(torch.float32)
+    for perturbation in self._perturbations:
+      audio = perturbation(audio, sample_rate, random_state)
+    audio = audio.numpy().squeeze()
+    # output_shape = audio.shape
+    # print("change shape from {} to {}".format(input_shape, output_shape))
+    assert isinstance(audio, np.ndarray)
+    assert len(audio.shape) == 1
+    return audio
+
+  def sox(self, audio, sample_rate, random_state, sox_effects):
+    sox_effects = sox_effects or []
+    speed = False
+    if self._speed is not None:
+      if random_state.random() < self._speed.prob:
+        factor = random_state.random() * (self._speed.max - self._speed.min) + self._speed.min
+        sox_effects.append(["speed", str(factor)])
+        sox_effects.append(["rate", str(sample_rate)])
+        speed = True
+    if self._tempo is not None:
+      if random_state.random() < self._tempo.prob and not speed:
+        factor = random_state.random() * (self._tempo.max - self._tempo.min) + self._tempo.min
+        sox_effects.append(["tempo", str(factor)])
```

Comment:
Right now this seems a bit strange to me. This function might be called multiple times and each time, the `sox_effects` will be appended, right?